### PR TITLE
Update FreeBSD native version

### PR DESCRIPTION
**Pull request by lapo-luchini**
Based on ab3d037ef586985bc1fa8601fcaa51154d3bfce0.
Built on: FreeBSD 12.1-RELEASE-p5 GENERIC  amd64.
Tests run: 298, Failures: 0, Errors: 0, Skipped: 2.

### CODE REVIEW

**Comment by michael-o** [comment]
I think if one uses binary versions, they should be built with the oldest available FreeBSD version, e.g., 11.3-RELEASE. Have already seen issues with 12 build binaries for TeX Live which did not work properly (linker issues).

**Comment by lapo-luchini** [comment]
Curiously enough the single test that fails after build on FreeBSD-11.3 (both i386 and amd64) runs perfectly when that binary is ran on FreeBSD-12.1/amd64.

**Comment by michael-o** [comment]
This needs to be checked. It maybe some uncertainly in the test. Can you post the failure details?

**Comment by lapo-luchini** [comment]
Here it is: (from the amd64 build, but it's the same on the i386 one)
[ErrorMessageTest.txt](https://github.com/xerial/sqlite-jdbc/files/4773000/ErrorMessageTest.txt)

**Comment by lapo-luchini** [comment]
Mhhh… now I notice this on FreeBSD 11.3:
`Tests run: 298, Failures: 1, Errors: 0, Skipped: 2.`
…which is _one more test_ than it is run on FreeBSD 12.1:
`Tests run: 298, Failures: 0, Errors: 0, Skipped: 2.`

**Comment by lapo-luchini** [comment]
Both 11.3 and 12.1 are running:
`Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)`
While OpenJDK8 version is slightly different: `1.8.0_242` vs `1.8.0_252`.

**Comment by michael-o** [comment]
Both say 298 tests. Are you certain?

**Comment by lapo-luchini** [comment]
Err, sorry, I read 298 as "passed", not as "total".

**Comment by michael-o** [comment]
I cannot reproduce this with `gmake native` and `mvn test`:

```
FreeBSD mika-ion 11.3-RELEASE-p7 FreeBSD 11.3-RELEASE-p7 #0: Tue Mar 17 08:12:57 UTC 2020     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  i386
```
and
```
FreeBSD deblndw011x.ad001.siemens.net 12.1-STABLE FreeBSD 12.1-STABLE #2 r361027: Thu May 14 16:04:27 CEST 2020     root@deblndw011x.ad001.siemens.net:/usr/obj/usr/src/amd64.amd64/sys/DEBLNDW011X  amd64
```

I have used the SQLite version downloaded via curl. Will try external amalgamation.

**Comment by lapo-luchini** [comment]
I ran that on VirtualBox using the official vmdk images for [amd64](https://download.freebsd.org/ftp/releases/VM-IMAGES/11.3-RELEASE/amd64/Latest/FreeBSD-11.3-RELEASE-amd64.vmdk.xz) and[ i386](https://download.freebsd.org/ftp/releases/VM-IMAGES/11.3-RELEASE/i386/Latest/FreeBSD-11.3-RELEASE-i386.vmdk.xz).
I simply (as far as I can remember):
```
# pkg install openjdk8 mvn curl bash gmake
# gmake native
# mvn test
```
Uh, one difference is that on those virtual machines I didn't bother creating an user account, so I ran it all as root, while the 12.1 box in my house factotum-server and I use a proper user there.

**Comment by lapo-luchini** [comment]
Indeed, running `mvn test` as a normal user returns zero errors on 11.3/amd64 too!

**Comment by michael-o** [comment]
Did more tests on 11.3 with external amalgamation from LOCALBASE, a lot of tests fail. I know why: when I started porting this driver to HP-UX for some applications. Internal amalagmation sets a lot of defines, I started to strip them down. This is the miminal set of config required for SQLite3:
```
CPPFLAGS="-DSQLITE_HAVE_ISNAN -DSQLITE_ENABLE_COLUMN_METADATA"
LDFLAGS="-Wl,+concatrpath -Wl,+b -Wl,$LIBDIR -L$LIBDIR" $CONFIGURE --enable-readline
gmake install
```

These are unclear:
```
#         -DSQLITE_ENABLE_FTS3
#         -DSQLITE_ENABLE_FTS3_PARENTHESIS
#         -DSQLITE_ENABLE_STAT2
#         -DSQLITE_DEFAULT_MEMSTATUS=0
#         -DSQLITE_MAX_MMAP_SIZE=1099511627776
```

But since the FreeBSD ports has a lot of options, tests fail because the drivers rely on those.

The task is: figure out which features are crucial for the driver and which tests need to be conditional.

**Comment by michael-o** [comment]
I have good news, my external amalgamation change makes life easier for you:
```
$ gmake native SQLITE_OBJ=/usr/local/lib/libsqlite3.so SQLITE_HEADER=/usr/local/include/sqlite3.h
"$JAVA_HOME/bin/javac"  -d target/common-lib -sourcepath src/main/java -h target/common-lib src/main/java/org/sqlite/core/NativeDB.java
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: src/main/java/org/sqlite/jdbc3/JDBC3Connection.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
mv target/common-lib/org_sqlite_core_NativeDB.h target/common-lib/NativeDB.h
cc -Itarget/sqlite-3.31.1-FreeBSD-x86_64 -I/usr/local/include -I/usr/local/openjdk8//include -Ilib/inc_linux -Os -fPIC -fvisibility=hidden   -I"/usr/local/openjdk8/include" -I"/usr/local/openjdk8/include/freebsd" -I target/common-lib -c -o target/sqlite-3.31.1-FreeBSD-x86_64/NativeDB.o src/main/java/org/sqlite/core/NativeDB.c
cc -Itarget/sqlite-3.31.1-FreeBSD-x86_64 -I/usr/local/include -I/usr/local/openjdk8//include -Ilib/inc_linux -Os -fPIC -fvisibility=hidden   -I"/usr/local/openjdk8/include" -I"/usr/local/openjdk8/include/freebsd" -o target/sqlite-3.31.1-FreeBSD-x86_64/libsqlitejdbc.so target/sqlite-3.31.1-FreeBSD-x86_64/NativeDB.o /usr/local/lib/libsqlite3.so -shared
cp target/sqlite-3.31.1-FreeBSD-x86_64/libsqlitejdbc.so /tmp/libsqlitejdbc.so
strip /tmp/libsqlitejdbc.so
cp /tmp/libsqlitejdbc.so target/sqlite-3.31.1-FreeBSD-x86_64/libsqlitejdbc.so
cp target/sqlite-3.31.1-FreeBSD-x86_64/libsqlitejdbc.so src/main/resources/org/sqlite/native/FreeBSD/x86_64/libsqlitejdbc.so
cp target/sqlite-3.31.1-FreeBSD-x86_64/libsqlitejdbc.so target/classes/org/sqlite/native/FreeBSD/x86_64/libsqlitejdbc.so
osipovmi@deblndw011x:~/var/Projekte/sqlite-jdbc ((d0d82ce...) *)
$ ldd target/sqlite-3.31.1-FreeBSD-x86_64/libsqlitejdbc.so
target/sqlite-3.31.1-FreeBSD-x86_64/libsqlitejdbc.so:
        libsqlite3.so.0 => /usr/local/lib/libsqlite3.so.0 (0x800e00000)
        libc.so.7 => /lib/libc.so.7 (0x80024d000)
        libz.so.6 => /lib/libz.so.6 (0x800677000)
        libm.so.5 => /lib/libm.so.5 (0x800692000)
        libthr.so.3 => /lib/libthr.so.3 (0x8006c3000)
```

**Comment by lapo-luchini** [comment]
Nice!
BTW: I think that's related to #495 (updating the FreeBSD Ports), not to this PR to have updated libraries in the JAR. 😇

**Comment by michael-o** [comment]
> 
> 
> Nice!
> BTW: I think that's related to #495 (updating the FreeBSD Ports), not to this PR to have updated libraries in the JAR. 😇

Apologies, yes. I have too many tabs open. All tests shall pass if the custom compiled code is used.

**Comment by lapo-luchini** [comment]
Anyways… that one test seemed to be failing only when runinng `mvn test` as `root`, and (on the same virtual host) works perfectly after I created an unpriviledged user.
I also noticed that the version built on 11.3 and 12.1 produce an identical `ldd` signature (and I tested the library on an actual 12.1), so I'm pretty sure the `so`s in this PR's commit work on FreeBSD 11&12, i386&amd64.

**Comment by michael-o** [comment]
You should  try to recreate the test manually with the cli tool.

**Comment by lapo-luchini** [comment]
Yes, it's the test itself that its flawed when ran as `root`:

```
% sqlite3 test.db
SQLite version 3.31.1 2020-01-27 19:55:54
Enter ".help" for usage hints.
sqlite> create table sample(id, name);
sqlite> INSERT INTO sample VALUES (1, 'ciao');
sqlite> ^D
% chmod 444 test.db
% sqlite3 test.db
SQLite version 3.31.1 2020-01-27 19:55:54
Enter ".help" for usage hints.
sqlite> INSERT INTO sample VALUES (2, 'foo');
Error: attempt to write a readonly database
sqlite> ^D
% sudo sqlite3 test.db
SQLite version 3.31.1 2020-01-27 19:55:54
Enter ".help" for usage hints.
sqlite> INSERT INTO sample VALUES (2, 'foo');
sqlite> ^D
```
PS: I tried this on both FreeBSD 12.1 and CentOS Linux 7.5.

**Comment by michael-o** [comment]
Where you able to figure out why with truss?

**Comment by lapo-luchini** [comment]
Yes, SQLite "read only" feature leverages on filesystem access control: if the database file (or folder, I suppose) is read-only to the user, the DB is accessed in a read-only way. If `sqlite3` is executed as `root` it is not limited by the file being read-only at all, and thus the database is never read-only. That specific test checks that a read-only file is effectively a read-only database, and as such if gives a false positive (or rather, a true _but probably unintended_ positive).

Basically, there's no issue at all with the `.so` libraries in this PR, it is the test case that is at fault.
(and gives the same "problem" also on Linux, I checked)

**Comment by gotson** [comment]
@lapo-luchini do you think it would be possible to use a cross-compiler inside a docker image, like for the other native libraries?

Maybe via something like [that](https://github.com/MrDOS/freebsd-cross-build), [that](https://github.com/myfreeweb/docker-freebsd-cross) or [that](https://github.com/sandvine/freebsd-cross-build).

**Comment by michael-o** [comment]
All of them seem to be GCC while LLVM is the native toolchain. Does it link anything outside of base?

**Comment by gotson** [comment]
Any chance you could compile the native lib against the latest version? I'm using it in my fork and that would allow me to upgrade 🙂

**Comment by xerial** [comment]
We will use #639 so that we can build native libraries without using machines with FreeBSD

**Comment by michael-o** [comment]
I know this from Jansi. This is ancient FreeBSD 9 with an old GCC. I'd still prefer a port created.

**Comment by gotson** [comment]
> I know this from Jansi. This is ancient FreeBSD 9 with an old GCC. I'd still prefer a port created.

Maybe I don't understand, but to me a port is like a package (like apt on Linux), right?

In that case, it's not the purpose of this project to compile sqlite for freebsd as a port, no?

Or did I misunderstood something?

**Comment by michael-o** [comment]
> 
> 
> > I know this from Jansi. This is ancient FreeBSD 9 with an old GCC. I'd still prefer a port created.
> 
> Maybe I don't understand, but to me a port is like a package (like apt on Linux), right?

Not really, a port compiles software from sources for the exact platform and CPU, rather than installing binary version which might or might now beeen compiled for that exact OS version and CPU bits.
 
> In that case, it's not the purpose of this project to compile sqlite for freebsd as a port, no?
> 
> Or did I misunderstood something?

That would be nice, but generally one does not expect the project to provide a ports in the ports tree. Ideally, this would, but it is sufficient that the project is in a state which can be compiled w/o any modification.
